File: plugin/src/main/java/com/badbones69/crazycrates/v2/utils/quadcrates/QuadCrateSpiralHandler.java
package com.badbones69.crazycrates.v2.utils.quadcrates;

import com.badbones69.crazycrates.v2.utils.interfaces.SpiralControl;
import org.bukkit.Location;
import org.bukkit.World;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

public class QuadCrateSpiralHandler implements SpiralControl {

    @Override
    public @NotNull ArrayList<Location> getSpiralLocationClockwise(Location center) {
        World world = center.getWorld();
        double downwardsDistance = .05;
        double expandingRadius = .08;
        double y = center.getY();
        double radius = 0;
        int amount = 10;//Amount of particles in each circle
        int increaseRadius = 0;
        int nextLocation = 0;//The limit of how far the circle goes before reset.
        double increment = (2 * Math.PI) / amount;//Spacing
        ArrayList<Location> locations = new ArrayList<>();
        for (int i = 0; i < 60; i++) {
            double angle = nextLocation * increment;
            double x = center.getX() + (radius * Math.cos(angle));
            double z = center.getZ() + (radius * Math.sin(angle));
            locations.add(new Location(world, x, y, z));
            y -= downwardsDistance;
            nextLocation++;
            increaseRadius++;
            if (increaseRadius == 6) {
                increaseRadius = 0;
                radius += expandingRadius;
            }
            if (nextLocation == 10) {
                nextLocation = 0;
            }
        }
        return locations;
    }

    @Override
    public @NotNull ArrayList<Location> getSpiralLocationCounterClockwise(Location center) {
        World world = center.getWorld();
        double downwardsDistance = .05;
        double expandingRadius = .08;
        double y = center.getY();
        double radius = 0;
        int amount = 10;//Amount of particles in each circle
        int increaseRadius = 0;
        int nextLocation = 0;//The limit of how far the circle goes before reset.
        double increment = (2 * Math.PI) / amount;//Spacing
        ArrayList<Location> locations = new ArrayList<>();
        for (int i = 0; i < 60; i++) {
            double angle = nextLocation * increment;
            double x = center.getX() - (radius * Math.cos(angle));
            double z = center.getZ() - (radius * Math.sin(angle));
            locations.add(new Location(world, x, y, z));
            y -= downwardsDistance;
            nextLocation++;
            increaseRadius++;
            if (increaseRadius == 6) {
                increaseRadius = 0;
                radius += expandingRadius;
            }
            if (nextLocation == 10) {
                nextLocation = 0;
            }
        }
        return locations;
    }
}

File: plugin/src/main/java/com/badbones69/crazycrates/v2/utils/quadcrates/QuadCrateSpiralHandler.kt
package com.badbones69.crazycrates.v2.utils.quadcrates

import com.badbones69.crazycrates.v2.utils.interfaces.SpiralControl
import org.bukkit.Location
import kotlin.math.cos
import kotlin.math.sin

class QuadCrateSpiralHandler: SpiralControl {
    override fun getSpiralLocationClockwise(center: Location): ArrayList<Location> {
        val world = center.world
        val downwardsDistance = .05
        val expandingRadius = .08
        var y = center.y
        var radius = 0.0
        val amount = 10 //Amount of particles in each circle

        var increaseRadius = 0
        var nextLocation = 0 //The limit of how far the circle goes before reset.

        val increment = 2 * Math.PI / amount //Spacing

        val locations = java.util.ArrayList<Location>()
        for (i in 0..59) {
            val angle = nextLocation * increment
            val x = center.x + radius * cos(angle)
            val z = center.z + radius * sin(angle)
            locations.add(Location(world, x, y, z))
            y -= downwardsDistance
            nextLocation++
            increaseRadius++

            if (increaseRadius == 6) {
                increaseRadius = 0
                radius += expandingRadius
            }

            if (nextLocation == 10) {
                nextLocation = 0
            }
        }
        return locations
    }

    override fun getSpiralLocationCounterClockwise(center: Location): ArrayList<Location> {
        val world = center.world
        val downwardsDistance = .05
        val expandingRadius = .08
        var y = center.y
        var radius = 0.0
        val amount = 10 //Amount of particles in each circle

        var increaseRadius = 0
        var nextLocation = 0 //The limit of how far the circle goes before reset.

        val increment = 2 * Math.PI / amount //Spacing

        val locations = java.util.ArrayList<Location>()
        for (i in 0..59) {
            val angle = nextLocation * increment
            val x = center.x - radius * cos(angle)
            val z = center.z - radius * sin(angle)
            locations.add(Location(world, x, y, z))
            y -= downwardsDistance
            nextLocation++
            increaseRadius++

            if (increaseRadius == 6) {
                increaseRadius = 0
                radius += expandingRadius
            }

            if (nextLocation == 10) {
                nextLocation = 0
            }
        }
        return locations
    }
}